### ISSUE TITLE
Abstracting TYPE and INTERFACE

### ISSUE THREAD

**Issue by OneWingedShark** [Ada-Rapporteur-Group/UCI-Prototype#104]
> **NOTE:**  While the LRM has an "abstract-type" already, this isn't about _that_ —rather it's more about CS's Abstract Data Types— irrespective of this issue, new/altered RM terminology, as well as syntax, can be argued about after the model/concept is correct.

### !summary	
It is needful for maintainability that pre- and post-conditions, type-invarients, as well as other such annotations be collectable into an abstract form (i.e. “Abstract data-type” in CS parlance); also for the “shape” of a type (e.g. attributes, indexing, etc) and [therefore] its interface.

**TL;DR —** The intent here is to be able to (a) lift [possibly shared] proving annotations to a central point, cleaning up the interfaces for types in Ada-specs; and (b) to be able to "describe Ada's type-system within Ada's type-system".

### !issue	
The addition of pre-/post-conditions, type-invariants, and other such aspects have been a good thing, allowing verification & proof (SPARK), unfortunately, practical experience with the Ada.Containers type has shown that these can be very unwieldy, repetitive, and harm maintainability. —This is especially true of the aspects involving user-defined indexing.— It would therefore be prudent to have some sort of ABSTRACT TYPE whereby these aspect-definitions could reside, typically for the provider of libraries (rather than the consumer), and be used.

Note that abstract subtypes could also be useful: many data-structures have common constraints with subtly different or excepted subcomponents. (E.G. A tree with different/additional rules for its leaves than for interior nodes, or special rules for the root-node.)

Indexing in particular raises the ‘interfacing’ aspect, and motivates a separate, related need for doing something similar for the TYPE itself, describing its interface to the rest of the language — this, in turn, presents a unique opportunity: with a well-designed ABSTRACT INTERFACE construct allowing us to describe the “shape” of the type (e.g. indexing, attributes, discriminants, entries, etc), we could describe Ada’s type-system within the type-system. (This would allow for a good foundation for meta-programming, if it’s ever needed, as well as allowing the reification of the conceptual type-hierarchy presented in the LRM, as well as allowing a singular place to refer new programmers to more easily/quickly know which attributes belong to which category of types.)

For example: given an [generic] ABSTRACT INTERFACE describing a type indexed by an array of discrete types, we could then have ABSTRACT TYPES using said interface —say a B-Tree and a Trie— to present to the end-user/programmer an actual data-type whose objects can be indexed via [[Wide_]Wide_]String.

> **NOTE:** Correctly formulating the ABSTRACT INTERFACE, as described, should also solve [the co-derivation problem](https://github.com/Ada-Rapporteur-Group/User-Community-Input/issues/98#issue-2262580417).

As an additional example, and closer to the motivating incident, the children of Ada.Containers could have their pre-/post-conditions moved to corresponding ABSTRACT TYPES, and those factored out, similar to OOP.

Since ABSTRACT TYPES are more like “conceptual classes” and “Abstract Data Types” —and intended for usage in proof-systems— there are going to be internal and/or virtual ‘fields’ that will be relied upon; this gives us some motivation to import an idea from Delphi’s ‘PROPERTY’ notion: the ability to have these virtual-fields be implemented as functions (for reading), procedures (for writing), and/or renaming of object fields in the types that implement them. (Perhaps reusing the TYPE A_T IS RECORD … END RECORD notation, perhaps reusing/overloading the Read and Write aspects s.t. “COUNT : NATURAL WITH READ => FALSE” would be valid.)

**Comment by ARG-Editor** [comment]
Why are the existing interface types insufficient for this task? They
provide a way to "collect" contracts, profiles, and (some) aspects for a
type without a concrete implementation.
 
I ask, of course, because using an existing feature is always better than
creating a new one. And it seems to me that we already have done this (twice
in fact, since the original abstract types also serve this purpose).
 
           Randy.

**Comment by OneWingedShark** [comment]
As stated, the ABSTRACT TYPE & ABSTRACT INTERFACE herein is **not** about OOP in any measure—
The ABSTRACT TYPE, just a term/nothing to do with Ada/LRM's abstract type, is about lifting out the specification/interface and properties of the type (i.e. the Computer Science Abstract Data Type), such that:

1.  proofs upon those properties may be made,
2.  these may be checked/used within the `PRIVATE` portion, 
3.  there is nothing committing you to OOP, and 
4.  make the language/compiler do all the "dirty work" of bookkeeping these. 
– The thing being described is turning inside-out the static-polymorphism of `GENERICs`, where with generics you can create a construct that creates a type by-parts given an interface while providing a concrete implementation, this goes the other way: providing a concrete interface (and properties) for some not-yet-extant construct.

The ABSTRACT INTERFACE is, again just a term, nothing to do with Ada/LRM language, is about being able to specify Ada's type-system within the type-system: 
1. being able to specify indexing, certainly like `ARRAY`, but it could easily be extended to allow indexing a trie with a string, letting the character represent a node-path, it should subsume/Annex-J all the indexing aspects; 
2. it should be able to be inheritable, allowing the reification of the type-hierarchy; 
3. it should allow specification of attributes;
4. it should allow for describing discriminants; 
5. it should allow for the description of inherent operations [e.g. `Function "+"`]; 
6. the ability to specify that the item has `entries` , [i.e.. protected objects and tasks]. 
– In short, this is about making things easier for:
  - library-writers: indexing becomes much less convoluted,
  - new-to-Ada programmers:
    - this ![the conceptual type-hierarchy](https://upload.wikimedia.org/wikipedia/commons/thumb/c/c2/Ada_types.png/550px-Ada_types.png) can be described explicitly in-code,
    - consequently attributes are easily findable.
    - actually, since it's also [tangentially] about fixing indexing, it cleans up that mess and makes things more welcoming especially there.
  - compiler-writers, possibly, as the reification of the conceptual-type-system above could be leveraged by the compiler-writer,
  -  teachers: the reification od the conceptual-type-hierarchy makes things a little easier to explain to students.
  -  Possibly **enabling reflection and meta-programming**.

– in short, ABSTRACT INTERFACE is about the "shape" of the type, how it is used by the language, and how the programmer uses it.

**Comment by Lucretia** [comment]
Only read the first part so far, a few notes:

1) Sounds like a trait system.
2) meta-programming is ALWAYS needed for tooling, Java proved that. There are tools I would've written but I'm not writing part of an Ada compiler to do it properly.
3) UNICODE, UNICODE, UNICODE, UNICODE, UNICODE, UNICODE STRINGS!

**Comment by ARG-Editor** [comment]
> meta-programming is ALWAYS needed for tooling, Java proved that. There are tools I would've written
> but I'm not writing part of an Ada compiler to do it properly.

Sure, but that sort of thing is out of bounds for the Ada Standard (the Standard needs to stay focused on the programming language definition, lest it get too large for tools to process). Rather, there is a separate standard for tooling interfaces for Ada (ASIS). And more recently, there is LibAdaLang for that purpose. I could see developing a Standard around that.

> Unicode
ASIS uses Unicode strings, and I think LibAdaLang does too. Unicode for Ada has a separate topic here, let's not mix them!

**Comment by ARG-Editor** [comment]
> While the LRM has an "abstract-type" already, this isn't about that ...

Then please call it something else here!! This forum is about the Ada Language Standard, and it is confusing to have lots of talk using terminology that already has a specific meaning.

> is not about OOP ...

Fine, but neither is Ada's abstract types. You can use them to do OOP, of course, but they're more flexible that that.

> being able to specify indexing, certainly like ARRAY, but it could easily be extended to allow indexing a trie with a string,

You can of course index anything in Ada as it is currently defined, and you surely can index with a string. This just sounds like adding even more complication.

>it should allow specification of attributes;

This is a bad idea for a number of reasons, not the least of which is that using functions for the purpose of describing properties works just as well, is just as readable, and doesn't add any cognitive overhead (everybody understands functions).

>it should allow for describing discriminants;

Describing of them how??? What could you possibly need to describe that's not found already in the definition.

But the real problem here is that you have what appears to be a pie-in-the-sky idea that would take a massive amount of work to define. But you still have not explained anything that you could do with this in an Ada program (as opposed to an Ada tool) that you cannot do with the existing facilities. At least not anything that you NEED to do (as opposed to nice-to-have at best things).

**Comment by OneWingedShark** [comment]
>> meta-programming is ALWAYS needed for tooling, Java proved that. There are tools I would've written
but I'm not writing part of an Ada compiler to do it properly.
>
> Sure, but that sort of thing is out of bounds for the Ada Standard (the Standard needs to stay focused on the programming language definition, lest it get too large for tools to process).

Sure, but the meta-programming capabilities (e.g. reflection) certainly are _not_ out of the bounds of the standard.

>> While the LRM has an "abstract-type" already, this isn't about that ...
>
> Then please call it something else here!! This forum is about the Ada Language Standard, and it is confusing to have lots of talk using terminology that already has a specific meaning.

Alright...
Howabout a "property-store" for the construct which corresponds to the concept of abstract-data-type within the rest of C.S. — the abstracting and collecting of common properties whereby (e.g.) properties [of the type] are proven.

And "shape" for the thing that I've been describing as an abstract-interface: the thing which describes the type, abstractly, to include how to handle/interface with it.

>> being able to specify indexing, certainly like ARRAY, but it could easily be extended to allow indexing a trie with a string,
>
> You can of course index anything in Ada as it is currently defined, and you surely can index with a string. This just sounds like adding even more complication.

The current definition of user-defined-indexing is rather noisy, gets in the way, and overly confuses things in practice.
There are also some rather extreme limitations: you cannot use the `[ Key => Item ]` syntax for a constant object in a `preelaborate` unit because it depends on calls which violate the restrictions thereof. — I want to be able to use compile-time constants at the same level as the package dependencies, which is to say: if there were some magic `pure` container-unit, I should be able to say `Data : Magic_Container:= [ 1 => 1.1, 2 => 3.1, 3 => 4.2, 4 => 2.1 ];`, just as I can with an plain old `array` of `Float`.

>> it should allow specification of attributes;
>
> This is a bad idea for a number of reasons, not the least of which is that using functions for the purpose of describing properties works just as well, is just as readable, and doesn't add any cognitive overhead (everybody understands functions).

I disagree; not only does VHDL allow it, but newcomers are often perplexed as to _where_ attributes come from. It would be very nice to be able to point them to `STANDARD` and the meta-type-hierarchy that they would be used for. (i.e. Reification of the previously-posted image of the conceptual hierarchy of Ada's type-system) — It would also aid new compiler-writers in that this feature would allow the defining of the type-system within the type-system itself: allowing the constructs to be easily specifiable would allow leveraging the language to implement the language.

For example, one of my favorite "tricks" is using `Integer'Value( TEXT )` to parse integers just as Ada would/does; no need to go in and define some character-by-character method or RegEx or anything, just leverage the language!

Besides-which, even if we didn't want attribute-definitions willy-nilly, we could restrict them to this particular construct. (And/or, restrict "uptake" into the rest of the language to being within `STANDARD`, or `Ada.Typesystem` or something.)

> But the real problem here is that you have what appears to be a pie-in-the-sky idea that would take a massive amount of work to define.

I know it would.
The thing though is that we could allow Ada to define Ada more easily.
For example, what is the rule about eliminating the `.all` from an `access` [when unambiguous] but indexing with a null-sequence? (`Implicit_Dereference` does the same.)

> But you still have not explained anything that you could do with this in an Ada program (as opposed to an Ada tool) that you cannot do with the existing facilities. At least not anything that you NEED to do (as opposed to nice-to-have at best things).

I mean, the ability to meta-program is invaluable to tooling. (What we are talking about are two aspects of a meta-object typesystem description.)
The ability to tell a tool "_I'm looking for this **SHAPE**_" is pretty valuable, and that would have to operate on some method of describing the type-system.

Also, the ability to e.g. have compile-time constant containerized data would be very nice.

As would the ability to, say, add at the base metaobject default ASN.1 serialize/deserialize as the `'Input` and `'Output` stream-operation defaults.

**Comment by Lucretia** [comment]
> > meta-programming is ALWAYS needed for tooling, Java proved that. There are tools I would've written
> > but I'm not writing part of an Ada compiler to do it properly.
> 
> Sure, but that sort of thing is out of bounds for the Ada Standard (the Standard needs to stay focused on the programming language definition, lest it get too large for tools to process). Rather, there is a separate standard for tooling interfaces for Ada (ASIS). And more recently, there is LibAdaLang for that purpose. I could see developing a Standard around that.
> 
> > Unicode
> > ASIS uses Unicode strings, and I think LibAdaLang does too. Unicode for Ada has a separate topic here, let's not mix them!

What planet are you on? Of course IT IS! And if you hadn't noticed, ASIS is dead.

*  [Oberon-2 Reflection Model](http://norayr.am/papers/Moessenboeck%20-%20Oberon-2%20Reflection%20Model.pdf)
* [Java reflection]( https://www.oracle.com/technical-resources/articles/java/javareflection.html)
* [What languages have reflection](https://www.google.com/search?client=firefox-b-d&q=what+languages+have+reflection)

Your myopic stance on literally anything other than what you've already done is one of the reasons I started thinking about dropping Ada and writing my own language without all the baggage.

**Comment by Lucretia** [comment]
The ASN.1 stuff should be a library level stream package.

**Comment by OneWingedShark** [comment]
> The ASN.1 stuff should be a library level stream package.

Ah, I wasn't proposing we should add in ASN.1 — just that an implementation could hook it in so easily if there was a way to define the type system in terms of the type system, including attributes. Heck, it would be really easy to do it as implementation-defined attributes: just put the attributes `'Serialize` and `'Deserialize` at your base.

**Comment by Lucretia** [comment]
Again, you want a traits system combined with reflection which handles the language level meta stuff. But given how they complain about the complexity (above) and how they refuse to make the language simpler (other threads elsewhere), I can't see it happening.

**Comment by OneWingedShark** [comment]
> Again, you want a traits system combined with reflection which handles the language level meta stuff.
Sure.

Call it traits, call it SHAPEs, call it abstract interfaces... whatever.
It's the underlying structures, abstracting the way to interact with them. And it's _that_ which I'm proposing to construct a model of. (Notice I've been very careful not to propose any syntax, precisely because I want to get the model down correctly before we start bikeshedding.)

> But given how they complain about the complexity (above) and how they refuse to make the language simpler (other threads elsewhere), I can't see it happening.

That's just it!
If we abstract things out a little, we can use that abstraction to subsume the complexity.
Ada 2012's indexing capabilities are tied to the aspect system, and in hindsight was the wrong call: using the user-defined indexing is awkward, gets in the way, and inhibits readability. — If we had a good method for defining indexable types, as seamlessly as arrays/at the meta-level, we could Annex J the whole aspect-based system.

The pre-/post-conditions, type-invariants, etc are actually really well done, the problem is that for really big/complex data-types  like containers, we really need a way to keep them from overloading the source to the point where it becomes unreadable — arguably where things are w/ containers, in-particular — which is why we should also have a method of spinning those properties into a common location that can be referenced/incorporated, essentially what the wider Computer Science world terms _Abstract Data Types_, but could also be seen as "turning the static-polymorphism of generics inside-out."

**Comment by Lucretia** [comment]
But that;s just it, they won't annex j anything, they have invented it but refuse to use it.

**Comment by Lucretia** [comment]
I'm not sure if something like that would require macro support to implement properly.

**Comment by Lucretia** [comment]
I'll add what I added to irc earlier so it's logged somewhere:

```text
[11:05] <Lucretia> Shark8: https://www.reddit.com/r/ProgrammingLanguages/comments/v3j21i/what_meta_programming_approach_to_use_in_a_new/
[11:14] <Lucretia> Shark8: I definitely think a way to start this would be AST := Entity'AST
[11:14] <Lucretia> and with Ada.AST;
[11:15] <Lucretia> to get the AST node types.
[11:15] <Lucretia> and subprograms could be determined with work with the AST when they declare themselves as 'Meta.
[11:18] <Lucretia> Ada.AST.Attribute_Tables;
[11:19] <Lucretia> or Ada.Internals.AST and Ada.Internals.Attribute_Tables.
[11:19] <Lucretia> or Meta instead of Internals
[11:19] <Lucretia> shorter and matches with the other ideas
[11:20] <Lucretia> trait Compile_Time is new Attribute;
[11:21] <Lucretia> Attribute_Table.Add (Compile_Time);
[11:22] <Lucretia> would need a way to say what it applies to
[11:24] <Lucretia> Basically, it's raising the implementation up to the language level
[11:26] <Lucretia> but how would you add a function to the language without having to with it?
[11:35] <Lucretia> Shark8: you might want to grab Towards a formal description of Ada
```
